### PR TITLE
Revocation phase using durability policies

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

This PR adds the ability to check for revocation using the durability policies in `EmergencyReparentShard`. What that means is that instead of exiting incase of more than 1 failure, `ERS` will first try and find whether it is safe to proceed even when we have multiple failures! 
So, if the durability_policies allow for more than 1 failure, ERS will succeed!

Consider an example for the same. Let us say we are using `semi_sync` as our durability policy and we have four servers, A, B, C and D. A is our current primary, B and C are replicas and D is a rdonly tablet.
Lets say that A and D, both go down.

In the older version, ERS would have failed since there are more than 1 failures
But now, ERS will see that even with A and D down, we can proceed safely. Safety, here means that we can guarantee that the unreachable (equivalent to down) tablets cannot make forward progress. In this case, only A is eligible to be a primary and D cannot provide Semi_Sync ACKs, which means that A and D cannot make progress, so it is safe to continue with the ERS.

## Related Issue(s)

- #8975 

## Checklist
- [x] Should this PR be backported? No
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/test/endtoend/reparent/utils/utils.go
Comment (deepthi) [inline go/test/endtoend/reparent/utils/utils.go:176]: Can we replace this with `require.Fail` (or whatever the right equivalent is)?

File: go/vt/vtctl/reparentutil/durability.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/durability.go:2]: Until we decide what to do about the copyright, we should not be making any changes.
Technically, it is incorrect to change the copyright year.

File: go/vt/vtctl/reparentutil/durability_policies.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/durability_policies.go:153]: Why would you ever pass a PRIMARY as `replica` to this function?

File: go/vt/vtctl/reparentutil/durability_policies.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/durability_policies.go:180]: Don't we need this check in `durabilitySemiSync` as well?

File: go/vt/vtctl/reparentutil/durability_policies.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/durability_policies.go:184]: Same question as above about `PRIMARY`.

File: go/vt/vtctl/reparentutil/durability.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/durability.go:38]:
```suggestion
// RevokeForTablet checks whether we have reached enough tablets such that the given primary eligible tablet cannot accept any new writes
```

File: go/vt/vtctl/reparentutil/replication.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/replication.go:253]: You probably need to change this error message.

File: go/vt/vtctl/reparentutil/replication.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/replication.go:178]: What is the full timeout for ERS? what do we expect the typical value for `waitReplicasTimeout` to be?

**Review by deepthi — COMMENTED** [review]
Comments from a first pass. I still need to understand the new Revoke funcs.

File: go/vt/vtctl/reparentutil/durability.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability.go:2]: This is an unfortunate incorrect diff from GitHub...
I actually renamed the `durability.go` file to `durability_policies.go` since that file had the implementations of the durability policies.
Then later I add `durability.go` for having functions with respect to durability like `Revoked`.

So the file `durability_policies.go` has the year 2021, since it is unchanged except for rename. And `durability.go` now holds new code all introduced in this year.

File: go/vt/vtctl/reparentutil/durability_policies.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability_policies.go:153]: There are two reasons why that would happen - 
1. As a fail safe for, if we have multiple tablets thinking themselves to be primary
2. We want to know whether the current primary will be a sending semi sync acks for a replica (which is a potential primary at this point). So the `replica` argument can be a tablet of type Primary.

File: go/vt/vtctl/reparentutil/durability_policies.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability_policies.go:180]: I'll move that check to the Exported function which calls these methods, so that they can always assume that the provided values are non-negative

File: go/vt/vtctl/reparentutil/replication.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/replication.go:178]: Both are configurable, if you call ERS from the vtctlclient, then the default action_timeout (total ERS time) is 1 hour and `waitReplicasTimeout` is also configurable which defaults to 30 seconds.
However, if we want ERS to be able to proceed with more than 1 failures, we need the action_timeout to be more than wait_replicas_timeout.

File: go/test/endtoend/reparent/utils/utils.go
Comment (GuptaManan100) [reply go/test/endtoend/reparent/utils/utils.go:176]: Done

File: go/vt/vtctl/reparentutil/durability.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability.go:38]: Done

File: go/vt/vtctl/reparentutil/replication.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/replication.go]: Done

**Comment by GuptaManan100** [comment]
@deepthi I have renamed the files back so it is easier to review and added the requested comments in the `durabler` interface.

File: go/vt/vtctl/reparentutil/durability.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/durability.go:65]: Isn't this supposed to be higher, not lower?

File: go/vt/vtctl/reparentutil/durability.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/durability.go:67]:
```suggestion
	// semiSyncAckers represents the number of semi-sync ackers required for a given tablet if it were to become the PRIMARY instance
```

File: go/vt/vtctl/reparentutil/durability.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/durability.go:69]:
```suggestion
	// isReplicaSemiSync returns whether the "replica" should send semi-sync acks if "primary" were to become the PRIMARY instance
```

File: go/vt/vtctl/reparentutil/durability_funcs.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/durability_funcs.go:41]: Is this correct? Just because replication is stopped does not mean it can't accept writes, that is only true when semi-sync is enabled.

File: go/vt/vtctl/reparentutil/replication.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/replication.go:230]: How does this work with `ignoredTablets`? If we allow 1 failure but `ignoredTablets` is not empty, what happens?

File: go/vt/vtctl/reparentutil/durability.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability.go:65]: yes you are right, what was I thinking when I typed it

File: go/vt/vtctl/reparentutil/durability_funcs.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability_funcs.go:41]: For all tablets in this list that think they are Primary type, in `StopReplicationAndBuildStatusMaps`, we call `demotePrimary` on them, which succeeds only after setting the tablet to rdonly. So all the tablets in this list would not have just stopped replication but would also be read only, so just checking that the primary tablet is in the list is sufficient

File: go/vt/vtctl/reparentutil/durability.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability.go:65]: fixed

File: go/vt/vtctl/reparentutil/durability.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability.go:67]: fixed

File: go/vt/vtctl/reparentutil/durability.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability.go:69]: fixed

File: go/vt/vtctl/reparentutil/replication.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/replication.go:230]: IgnoredTablets functionality remains the same. We used to proceed with 1 failure in non-ignored tablets earlier which we will continue to do.
For the revocation check, we use the entire list of tablets to check for revocation to be sure that we only proceed with multiple failures if we can guarantee safety.

File: go/vt/topo/topoproto/tablet.go
Comment (ajm188) [inline go/vt/topo/topoproto/tablet.go:66]: Suggestion: rename to `IsTabletInList` to match `func IsTypeInList(tabletType topodatapb.TabletType, types []topodatapb.TabletType) bool {`

File: go/vt/topo/topoproto/tablet_test.go
Comment (ajm188) [inline go/vt/topo/topoproto/tablet_test.go:119]: suggestion -- run these test cases in parallel with:

```go
func TestBlahBlah(t *testing.T) {
    t.Parallel()

    for _, testcase := range testcases {
        testcase := testcase // prevent loop shadowing
        t.Run(testcase.name, func(t *testing.T) {
            t.Parallel()
            // rest of test
        })
    }
}
```

File: go/vt/topo/topoproto/tablet_test.go
Comment (ajm188) [inline go/vt/topo/topoproto/tablet_test.go:122]: I think you want `assert.Equal` here

File: go/vt/vtctl/reparentutil/durability_funcs.go
Comment (ajm188) [inline go/vt/vtctl/reparentutil/durability_funcs.go:39]: for these functions that check and return bools, i'd suggest renaming to from `<Thing>` to `Is<Thing>` to make them read a little nicer where they are used

File: go/vt/vtctl/reparentutil/durability_funcs.go
Comment (ajm188) [inline go/vt/vtctl/reparentutil/durability_funcs.go:61]: same naming suggestion here

**Review by ajm188 — COMMENTED** [review]
just some naming/testing suggestions! i think the rest is looking good :)

File: go/vt/vtctl/reparentutil/durability_funcs.go
Comment (deepthi) [reply go/vt/vtctl/reparentutil/durability_funcs.go:41]: Hmm.. so usage of this function is dependent on some pre-condition being satisfied.
Is there a reason it has to be a separate function in a separate file? If it is being used only once, it can as well be just a code block, and then you don't have to worry about whether callers are satisfying the pre-condition.
If it _MUST_ be in a separate file for some good reason, you should make these two `Revoke` functions unexpected, so that they cannot be used outside the package. `s/Revoke/revoke`.

Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability_funcs.go:41]: Yes that's a good point. The only reason that I have kept them as separate function is for testing and readability. It makes it much easier to have unit tests for these individual functions along with unit tests for the larger StopReplicationAndBuildStatusMap since that requires a larger setup with a tablet manager etc. 
Just to be extra careful, I'll add a comment as well there stating the assumption that the function has and unexport it

File: go/vt/vtctl/reparentutil/durability_funcs.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability_funcs.go:39]: How do you feel about, `haveRevokedForTablet`? instead of `isRevokedForTablet`?

File: go/vt/vtctl/reparentutil/durability_funcs.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability_funcs.go]: Fixed

File: go/vt/topo/topoproto/tablet.go
Comment (GuptaManan100) [reply go/vt/topo/topoproto/tablet.go]: Fixed

File: go/vt/topo/topoproto/tablet_test.go
Comment (GuptaManan100) [reply go/vt/topo/topoproto/tablet_test.go:120]: Fixed

File: go/vt/topo/topoproto/tablet_test.go
Comment (GuptaManan100) [reply go/vt/topo/topoproto/tablet_test.go]: Fixed

File: go/vt/vtctl/reparentutil/durability_funcs.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability_funcs.go]: I have made the change for `haveRevokedForTablet`. If you feel the other name is better, I'll rename it

File: go/vt/vtctl/reparentutil/durability_funcs.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/durability_funcs.go]: Changed it to `haveRevoked`

**Comment by GuptaManan100** [comment]
Thankyou @deepthi  and @ajm188  for the reviews! I have addressed all the review comments, could you take a look once, again?

File: go/vt/vtctl/reparentutil/durability_funcs.go
Comment (ajm188) [reply go/vt/vtctl/reparentutil/durability_funcs.go]: fine by me! as long as it implies the boolean/question nature, and we're consistent with what we use for both, i'm happy :)

**Comment by ajm188** [comment]
lgtm, but let's get deepthi's approval (and clean up the commit history length) before merging

**Review by deepthi — APPROVED** [review]

**Comment by deepthi** [comment]
> lgtm, but let's get deepthi's approval (and clean up the commit history length) before merging

Do you mean squash some of the commits together?
If so, I recommend squashing only the last 5 (since we started review) either into one, or into whatever is the most relevant older commit. Typically we don't squash at all (though I'm well aware that this makes certain things like reverts and bisects that much harder).

**Pull request merged**
